pkg/core/metrics.go

```diff
@@ -0,0 +1,245 @@
+	otherCount                = HeartbeatBreakdownHandleCount.WithLabelValues("Other")
+)
+
+func init() {
+	prometheus.MustRegister(HeartbeatBreakdownHandleDurationSum)
+	prometheus.MustRegister(HeartbeatBreakdownHandleCount)
+	prometheus.MustRegister(AcquireRegionsLockWaitDurationSum)
+	prometheus.MustRegister(AcquireRegionsLockWaitCount)
+}
+
+type saveCacheStats struct {
+	startTime              time.Time
+	lastCheckTime          time.Time
+	checkOverlapsDuration  time.Duration
+	validateRegionDuration time.Duration
+	setRegionDuration      time.Duration
+	updateSubTreeDuration  time.Duration
+}
+
+// RegionHeartbeatProcessTracer is used to trace the process of handling region heartbeat.
+type RegionHeartbeatProcessTracer interface {
+	Begin()
+	OnPreCheckFinished()
+	OnAsyncHotStatsFinished()
+	OnRegionGuideFinished()
+	OnSaveCacheBegin()
+	OnSaveCacheFinished()
+	OnCheckOverlapsFinished()
+	OnValidateRegionFinished()
+	OnSetRegionFinished()
+	OnUpdateSubTreeFinished()
+	OnAllStageFinished()
+	LogFields() []zap.Field
+}
+
+type noopHeartbeatProcessTracer struct{}
+
+// NewNoopHeartbeatProcessTracer returns a noop heartbeat process tracer.
+func NewNoopHeartbeatProcessTracer() RegionHeartbeatProcessTracer {
+	return &noopHeartbeatProcessTracer{}
+}
+
+func (n *noopHeartbeatProcessTracer) Begin()                    {}
+func (n *noopHeartbeatProcessTracer) OnPreCheckFinished()       {}
+func (n *noopHeartbeatProcessTracer) OnAsyncHotStatsFinished()  {}
+func (n *noopHeartbeatProcessTracer) OnRegionGuideFinished()    {}
+func (n *noopHeartbeatProcessTracer) OnSaveCacheBegin()         {}
+func (n *noopHeartbeatProcessTracer) OnSaveCacheFinished()      {}
+func (n *noopHeartbeatProcessTracer) OnCheckOverlapsFinished()  {}
+func (n *noopHeartbeatProcessTracer) OnValidateRegionFinished() {}
+func (n *noopHeartbeatProcessTracer) OnSetRegionFinished()      {}
+func (n *noopHeartbeatProcessTracer) OnUpdateSubTreeFinished()  {}
+func (n *noopHeartbeatProcessTracer) OnAllStageFinished()       {}
+func (n *noopHeartbeatProcessTracer) LogFields() []zap.Field {
+	return nil
+}
+
+type regionHeartbeatProcessTracer struct {
+	startTime             time.Time
+	lastCheckTime         time.Time
+	preCheckDuration      time.Duration
+	asyncHotStatsDuration time.Duration
+	regionGuideDuration   time.Duration
+	saveCacheStats        saveCacheStats
+	OtherDuration         time.Duration
+}
+
+// NewHeartbeatProcessTracer returns a heartbeat process tracer.
+func NewHeartbeatProcessTracer() RegionHeartbeatProcessTracer {
+	return &regionHeartbeatProcessTracer{}
+}
+
+func (h *regionHeartbeatProcessTracer) Begin() {
+	now := time.Now()
+	h.startTime = now
+	h.lastCheckTime = now
+}
+
+func (h *regionHeartbeatProcessTracer) OnPreCheckFinished() {
+	now := time.Now()
+	h.preCheckDuration = now.Sub(h.lastCheckTime)
+	h.lastCheckTime = now
+	preCheckDurationSum.Add(h.preCheckDuration.Seconds())
+	preCheckCount.Inc()
```

Comment:
Is it possible to flush these metrics asynchronously? It appears that setting the metrics in the middle of processing may affect the performance.

Reply:
I think it's ok because all of it use atomic. and we can dynamically disable the trace by the config `EnableHeartbeatBreakdownMetrics`